### DISCUSSION TITLE
Usage of global data?

### DISCUSSION THREAD

**Discussion by lucianoiam** [wasmerio/wasmer#2791] [Q&A]
Does Wasmer  (or the underlying Rust innards I don't know anything about!)  makes use of global / static data and/or singletons?

I'd like to understand the safety of embedding wasmer in a plugin context as I've done [here](https://github.com/lucianoiam/hiphop) to better document my implementation, specifically what are the known limitations.

**Comment by SamuraiCrow** [comment]
I'm not sure if this will help you but [WASI-libC](https://github.com/WebAssembly/wasi-libc) seems to have a single threaded implementation of libC.  When linked as multithreaded (linked below) it seems better than the one with WAMR-sdk.  I'm working on some changes to add thread-safety to the implementation linked here with a few mutexes.  My multithreaded fork is [here](https://github.com/SamuraiCrow/wasi-libc) if you want to try it sooner.  The one that works multithreaded is in the posix branch.

**Comment by lucianoiam** [comment]
Thanks, that looks interesting. Actually WASI hasn't a critical place in my project, it is focused in running basic WebAssembly modules compiled from AssemblyScript mostly for doing DSP math. Everything else is expected to be done in C++ for now.

In a plugin context it is desirable to avoid global data and in the specific audio plugin case avoid locking and syscalls from within the realtime thread.

Great to see how the WASM ecosystem is expanding !
